### PR TITLE
BREAKING CHANGE: Makes Job Parameters immutable.

### PR DESCRIPTION
**Pull request by andyHa**
As we more and more tend to use parameters as globally shared constants, it is neccesarry to make them immutable once shared.

Therefore we differentiate between "ParameterBuilder", which is the mutable way of setting up a parameter as wanted and "Parameter" itself, which is then an immutable object which can safely be shared.

This leads to some refactorings and simplifications, hence the amount of changed files. The actual changes are quite small.

### CODE REVIEW

File: src/main/java/sirius/biz/jobs/params/Parameter.java [hunk 1]
```diff
@@ -9,135 +9,32 @@
...
 import sirius.kernel.commons.Value;
-import sirius.kernel.health.Exceptions;
+import sirius.kernel.di.transformers.Composable;
 import sirius.kernel.nls.NLS;
 
 import java.util.Map;
 import java.util.Optional;
 
 /**
  * Defines a parameter which can be rendered in the UI and verify and extract a value from a context.
```
Comment (jakobvogel) [inline src/main/java/sirius/biz/jobs/params/Parameter.java:19]: Who can verify and extract? The parameter? (Clarify?)

File: src/main/java/sirius/biz/jobs/params/ParameterBuilder.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import sirius.kernel.health.Exceptions;
+import sirius.kernel.nls.NLS;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Provides a mutable instance which can be used to build {@link Parameter parameters}.
+ * <p>
+ * A parameter is setup using sublasses of this class which are eventually converted using {@link #build()}.
```
Comment (jakobvogel) [inline src/main/java/sirius/biz/jobs/params/ParameterBuilder.java:21]:
```suggestion
 * A parameter is set up using subclasses of this class which are eventually converted using {@link #build()}.
```

File: src/main/java/sirius/biz/jobs/params/ParameterBuilder.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+    @SuppressWarnings("unchecked")
+    protected P self() {
+        return (P) this;
+    }
+
+    /**
+     * Creates an immutable parameter from this builder.
+     * <p>
+     * Once the parameter is build, the builder itself should no longer be used.
```
Comment (jakobvogel) [inline src/main/java/sirius/biz/jobs/params/ParameterBuilder.java:63]:
```suggestion
     * Once the parameter is built, the builder itself should no longer be used.
```

File: src/main/java/sirius/biz/jobs/params/ParameterBuilder.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        }
+    }
+
+    /**
+     * Checks and transforms the given value.
+     *
+     * @param input the input wrapped as <tt>Value</tt>
+     *              * @return a serialized string version of the given input which can later be resolved using
+     *              * {@link #resolveFromString(Value)}
+     * @return the value represented as string
```
Comment (jakobvogel) [inline src/main/java/sirius/biz/jobs/params/ParameterBuilder.java:197]: Which of the two `@return`?

**Review by jakobvogel — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/jobs/batch/file/ArchiveImportJob.java [hunk 5]
```diff
@@ -119,7 +123,7 @@ protected boolean containsEntries(String... fileNamesToCheck) {
 
     @Override
     protected final boolean canHandleFileExtension(@Nullable String fileExtension) {
-        return FILE_EXTENSION_ZIP.equalsIgnoreCase(fileExtension);
+        return "zip".equalsIgnoreCase(fileExtension);
```
[the remark this answers is not in the document]
Comment (andyHa) [reply src/main/java/sirius/biz/jobs/batch/file/ArchiveImportJob.java:126]: This is due to another refactoring concerning archive files - consider this state temporary and the next PR will hopefully fix this...

File: src/main/java/sirius/biz/jobs/batch/file/FileImportJob.java [hunk 6]
```diff
@@ -30,31 +33,50 @@
...
-        VirtualFile file = process.require(fileParameter);
+        VirtualFile file = process.require(FILE_PARAMETER);
 
         if (canHandleFileExtension(Value.of(file.fileExtension()).toLowerCase())) {
             try (FileHandle fileHandle = file.download()) {
                 backupInputFile(file.name(), fileHandle);
                 executeForSingleFile(file.name(), fileHandle);
             }
-        } else if (FILE_EXTENSION_ZIP.equalsIgnoreCase(file.fileExtension())) {
+        } else if ("zip".equalsIgnoreCase(file.fileExtension())) {
```
[the remark this answers is not in the document]
Comment (andyHa) [reply src/main/java/sirius/biz/jobs/batch/file/FileImportJob.java:79]: s.a.

**Review by jakobvogel — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Review by fhaScireum — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Pull request merged**